src/Tasks/Copy.cs

```diff
@@ -374,15 +429,137 @@ CopyFileWithState copyFile
+        /// Parallelize I/O with the same semantics as the single-threaded copy method above.
+        /// ResolveAssemblyReferences tends to generate longer and longer lists of files to send
+        /// to CopyTask as we get further and further down the dependency graph.
+        /// The OS can handle a lot of parallel I/O so let's minimize wall clock time to get
+        /// it all done.
+        /// </summary>
+        private bool CopyParallel(
+            CopyFileWithState copyFile,
+            int parallelism,
+            out List<ITaskItem> destinationFilesSuccessfullyCopied)
+        {
+            bool success = true;
+
+            // We must supply the same semantics as the single-threaded version above:
+            //
+            // - For copy operations in the list that have the same destination, we must
+            //   provide for in-order copy attempts that allow re-copying different files
+            //   and avoiding copies for later files that match SkipUnchangedFiles semantics.
+            //   We must also add a destination file copy item for each attempt.
+            // - The order of entries in destinationFilesSuccessfullyCopied must match
+            //   the order of entries passed in, along with copied metadata.
+            // - Metadata must not be copied to destination item if the copy operation failed.
+            //
+            // We split the work into different Tasks:
+            //
+            // - Entries with unique destination file paths each get their own parallel operation.
+            // - Each subset of copies into the same destination get their own Task to run
+            //   the single-threaded logic in order.
+            //
+            // At the end we reassemble the result list in the same order as was passed in.
+
+            // Map: Destination path -> indexes in SourceFiles/DestinationItems array indices (ordered low->high).
+            var partitionsByDestination = new Dictionary<string, List<int>>(
+                DestinationFiles.Length, // Set length to common case of 1:1 source->dest.
+                StringComparer.OrdinalIgnoreCase);
+
+            for (int i = 0; i < SourceFiles.Length && !_canceling; ++i)
+            {
+                ITaskItem destItem = DestinationFiles[i];
+                string destPath = destItem.ItemSpec;
+                if (!partitionsByDestination.TryGetValue(destPath, out List<int> sourceIndices))
+                {
+                    // Use 1 for list length - common case is for no destination overlap.
+                    sourceIndices = new List<int>(1);
+                    partitionsByDestination[destPath] = sourceIndices;
+                }
+                sourceIndices.Add(i);
+            }
+
+            var successFlags = new IntPtr[DestinationFiles.Length];  // Lockless flags updated from each thread.
+            var actionBlockOptions = new ExecutionDataflowBlockOptions { MaxDegreeOfParallelism = parallelism };
```

Comment:
Can you plumb cancellation through here too? I think you should switch `_canceling` to be a `CancellationTokenSource`, then pass that through here. Jeff did something similar in https://github.com/Microsoft/msbuild/pull/3291/commits/4501ce63d7fe623021fef602943eb4b21c40a2bb.